### ISSUE TITLE
Imperfect cropping or BoundingBox calculation

### ISSUE THREAD

**Issue by seisman** [GenericMappingTools/gmt#4302]
**Description of the problem**

See `doc/scripts/GMT_-B_time4.ps` for an example. 

Here is the screenshot of the PS file:

![image](https://user-images.githubusercontent.com/3974108/95514509-0e00c380-098a-11eb-88d9-f564a3490bc1.png)

Here is the screenshot of the PDF file generated by `gmt psconvert -A -P -Tf xxx.ps`:
![image](https://user-images.githubusercontent.com/3974108/95514589-312b7300-098a-11eb-9dbd-085987d32b40.png)

FYI, I'm using gs 9.53.3 on macOS.

**Comment by PaulWessel** [comment]
We can rule out Preview as it looks exactly the same in Adobe Acrobat, so it is done via psconvert (and possibly gs).  I created the EPS first and find

```
%%BoundingBox: 0 0 343 27
%%HiResBoundingBox: 0 0 342.2612 26.4586
```

we then use the HiResBoundingBox in determining the final box for cropping.  Appending **-S** to your command says **-g**3423x265 **-r**720, i.e., default PDF image dpi (not relevant for the vector graphics) is 720 dpi but the window size is 3423 by 265.  Note these seems to be in 1/10th of a point and is rounded up, but it may still be affected by whatever (x0,y0) round-off takes place.  The usage messages talks about DEVICEWIDTH and HEIGHT.  Looking at the gs manual they list these -d options:

```
-dDEVICEWIDTHPOINTS=w 
-dDEVICEHEIGHTPOINTS=h
Sets the initial page width to w or initial page height to h respectively, specified in 1/72" units.
```

Just using those options with the same args give the same plot.  However, I notice in the PDF we get that there is some white space on top above the frame, as well as to both the left and right, so this seems to be a rounded-off y-shift?  I wonder if we introduce some shift when we take the raw BB values and compute w = x2-x1 and h = y2-y1 and then use 0,0 as the start.  Is 0,0 the baseline and curved letters can hang below?  Any thoughts, @joa-quim ?

**Comment by PaulWessel** [comment]
Hm, some more thoughts.  Forget about my 1/10th above.  Here is what happens.  We have

```
pix_w = urint (psconvert_smart_ceil (w * Ctrl->E.dpi / 72.0));
pix_h = urint (psconvert_smart_ceil (h * Ctrl->E.dpi / 72.0));
```

and these are used to set **-g**.  But what does E.dpi have to do with PDF? I think this is the problem.  Here it is 720 hence 720/72 gives my factor of 10, but 720 is only used for internal images.  I think we may need to switch to -dDEVICEWIDTHPOINTS=342.2612 -dDEVICEHEIGHTPOINTS=26.4586 unless the -g can take decimals.

**Comment by PaulWessel** [comment]
So **-g** takes pixels, hence integers, but PDF is not pixelled. Hi @joa-quim, perhaps we need an entirely different option for this like the EPScrop and other settings.  Surely, pixel counts is not what is required here.

**Comment by PaulWessel** [comment]
Pinging @joa-quim on this issue.

**Comment by PaulWessel** [comment]
Pinging @joa-quim one more time.

**Comment by PaulWessel** [comment]
So, some updates: This problem is of our own making it seems.  I have verified that it is related to our use of integer _PostScript_ coordinates using a dpi of 1200:

`#define PSL_DOTS_PER_INCH	1200`

If I temporarily change that to 12000 and rebuild GMT then the crop problem above goes away.  It is possibly related to a command that translates a new origin but uses integer units and is thus possibly off a tiny fraction.  If we can identify that one and adjust we may still be able to stick with our 1200 scheme.  However, differencing the PS and the EPS file only shows teh updated BoundingBox statements and a single floating point translate command with enough precision, so it is not so clear what to do.  The alternative is to increase that DPI and get a bit larger PS files but with much higher precision.  Thoughts on this?

**Comment by PaulWessel** [comment]
Oh, answer my own question: The coordinates we use to place the text (say 6 am) are in integer form and thus only accurate to nearest 1/1200 inch. If we extend the BB and the translate to the next full multiple of 1/1200 then I think we will be out of the woods.  This may in some cases lead to tiny white space between bounding box and 6 am, which presumably is better than a tiny bit of 6 am being clipped.  Is this an OK plan, @joa-quim and @seisman ?

**Comment by joa-quim** [comment]
Yep, a little invisible white margin is preferable to its clipping counterpart.

**Comment by seisman** [comment]
>  If we extend the BB and the translate to the next full multiple of 1/1200 then I think we will be out of the woods. This may in some cases lead to tiny white space between bounding box and 6 am, which presumably is better than a tiny bit of 6 am being clipped

Yes, sounds a good solution.

**Comment by PaulWessel** [comment]
Initial attempts not yet working.  I now take the reported bounding box (with 6 decimals from `gs -q -dNOSAFER -dNOPAUSE -dBATCH -sDEVICE=bbox example.ps`):

`%%HiResBoundingBox: 70.939404 46.601999 413.200605 73.060592`

and adjust like this (precision = 1/1200):

```
*x0 = floor (*x0 / precision) * precision;
*y0 = floor (*y0 / precision) * precision;
*x1 = ceil  (*x1 / precision) * precision;
*y1 = ceil  (*y1 / precision) * precision;
```

and that gives a slightly larger BB and dimensions:

```
Adjusted HighResBoundingBox from 70.939404/46.601999/413.200605/73.060592 to 70.939167/46.601667/413.200833/73.060833
Adjusted Dimensions from 342.261201 by 26.458593 to 342.261667 by 26.459167
```

and the origin is then shifted by **psconvert** thus:

`-70.939167 -46.601667 translate`

Yet, when I make the EPS first then manually try 

`gs -q -dNOPAUSE -dBATCH -dNOSAFER -dPDFSETTINGS=/prepress -dUseFlateCompression=true -dSCANCONVERTERTYPE=2 -sDEVICE=pdfwrite -dEPSCrop -dDEVICEWIDTHPOINTS=342.261667 -dDEVICEHEIGHTPOINTS=26.459167 -sOutputFile=example.pdf example.eps`

it gives the same cropped PDF.  The placement of  text at x,y is rounded to integers via

	return (PSL->internal.x0 + (int)lrint (x * PSL->internal.x2ix));

where = PSL->internal.x2ix = PSL_DOTS_PER_INCH.  So seems to me at most, x would be off by ±0.5 "dot", or 0.5/1200 points.  So not sure if my BB above could still be off? But it should reflect the bias in that coordinate anyway.   Now, 1/1200 points is

0.000833333333333

while with 6 decimals, it is 0.000833.  In terms of dots, that difference is just 0.04% so that cannot be the mismatch.

If I manually edit the translate and heights I can get it to fit OK but I have no way of tieing that back to my initial values.  Suggestions or comments or hints, please, @joa-quim and @seisman

**Comment by PaulWessel** [comment]
On the other side, I tried setting PSL_DOTS_PER_INCH tp 7200, meaning 1 integer PS init is 0.01 point (instead of our current 0.06 point). THen the problem goes away again.  You may recall the ugly alignment of letters along contours (ex25) that I solved by forcing an extra decimal in the output instead of using integers.  We could probably revert that if we used 7200.  Upside: Higher precision in positioning of items, like here.  Cons: Postscript files almost all change, but size only by 5-10%.  Need feedback on this.

**Comment by PaulWessel** [comment]
Note, in this configuration, the crop is ~perfect for the south, but we still have white space along EWN.  It may be a gs limitation as well here.

**Comment by joa-quim** [comment]
I don't think we should increase the size of all PS files because of this barely visible issue. And there is always -A1 if the clipping is too annoying.

**Comment by PaulWessel** [comment]
I can try a bit harder to see if I can get a better offset/size with the current values.  I just cannot get a one-to-one correspondence between what I do and what it results in, making it a bit frustrating.
-A1p is there, but for anything automatic it is a major hassle to have to do that.
Also, I think we need **psconvert** to use 

`-dDEVICEWIDTHPOINTS=w -dDEVICEHEIGHTPOINTS=h`

in floating-point points instead of 

`-g3423x267 -r720`

which is pixel based and suitable for making the PNGs etc.
